### PR TITLE
Move all solar calendars that use book arithmetic to calendrical_calculations

### PR DESCRIPTION
**Pull request by Manishearth**
Progress on #3975

I changed the way the I32Result thing worked so that I can use the `?` operator.

### CODE REVIEW

File: utils/calendrical_calculations/src/helpers.rs [hunk 1]
```diff
@@ -618,38 +618,39 @@ fn test_ceil_div() {
...
-            I32Result::AboveMax(_) => i32::MAX,
+            I32CastError::BelowMin => i32::MIN,
+            I32CastError::AboveMax => i32::MAX,
         }
     }
 }
 
 #[inline]
-pub const fn i64_to_i32(input: i64) -> I32Result {
+pub const fn i64_to_i32(input: i64) -> Result<i32, I32CastError> {
```
Comment (echeran) [inline utils/calendrical_calculations/src/helpers.rs:641]: praise: I like this improvement

File: components/calendar/src/persian.rs [hunk 2]
```diff
@@ -614,18 +542,6 @@ mod tests {
         }
     }
 
-    #[test]
```
Comment (echeran) [inline components/calendar/src/persian.rs:617]: Is there a reason why we don't preserve this test?

**Review by echeran — DISMISSED** [review]
LGTM, modulo a question about a test. Otherwise, seems straightforward, like the combination of 1) replacing `I32Result` with a proper `Result` and 2) moving over more calendar book logic into the new crate for that.

File: components/calendar/src/persian.rs (hunk 2 above)
Comment (Manishearth) [reply components/calendar/src/persian.rs:617]: the function it is testing no longer exists and the replacement function (persian_from_fixed) is already tested below

**Comment by Manishearth** [comment]
@robertbastian mind stamping this PR? autosubmit=true

was really hoping these PRs would not stretch over the long weekend

**Review by robertbastian — APPROVED** [review]

**Pull request merged**